src/main/java/net/snowflake/client/jdbc/cloud/storage/GcmEncryptionProvider.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (c) 2012-2019 Snowflake Computing Inc. All rights reserved.
+ */
+package net.snowflake.client.jdbc.cloud.storage;
+
+import static java.nio.file.StandardOpenOption.CREATE;
+import static java.nio.file.StandardOpenOption.READ;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.Base64;
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.CipherInputStream;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import net.snowflake.client.jdbc.MatDesc;
+import net.snowflake.common.core.RemoteStoreFileEncryptionMaterial;
+
+public class GcmEncryptionProvider {
+  private static final int TAG_LENGTH = 128;
+  private static final String AES = "AES";
+  private static final String FILE_CIPHER = "AES/GCM/NoPadding";
+  private static final String KEY_CIPHER = "AES/GCM/NoPadding";
+  private static final int BUFFER_SIZE = 8 * 1024 * 1024; // 2 MB
+  private static final int blockSize;
+  private static final SecureRandom random;
+
+  static {
+    try {
+      Cipher fileCipher = Cipher.getInstance(FILE_CIPHER);
+      blockSize = fileCipher.getBlockSize();
+
+      random = SecureRandom.getInstance("SHA1PRNG");
+      byte[] bytes = new byte[10];
+      random.nextBytes(bytes);
+    } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
+      throw new ExceptionInInitializerError(e);
+    }
+  }
+
+  public static InputStream encrypt(
+      StorageObjectMetadata meta,
+      long originalContentLength,
+      InputStream src,
+      RemoteStoreFileEncryptionMaterial encMat,
+      SnowflakeStorageClient client,
+      byte[] cekAad,
+      byte[] kekAad) // TODO GCM is it ok or should it be a part of meta?
+      throws InvalidKeyException, InvalidAlgorithmParameterException, IllegalBlockSizeException,
```

Comment:
do we need all exceptions to be reported in throws or can we use the base one?

Reply:
I'll let Kevin to decide.